tools/SourceKit/lib/SwiftLang/SwiftSignatureHelp.cpp

```diff
@@ -0,0 +1,270 @@
+#include "swift/AST/Decl.h"
+#include "swift/AST/Types.h"
+#include "swift/IDE/CodeCompletionResultPrinter.h"
+#include "swift/IDE/CodeCompletionStringBuilder.h"
+#include "swift/IDE/CommentConversion.h"
+#include "swift/IDE/SignatureHelp.h"
+#include "swift/IDETool/IDEInspectionInstance.h"
+#include "llvm/Support/Allocator.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace SourceKit;
+using namespace swift;
+using namespace ide;
+
+using ChunkKind = CodeCompletionString::Chunk::ChunkKind;
+
+namespace {
+struct SignatureInfo {
+  StringRef Text;
+  StringRef DocComment;
+  std::optional<unsigned> ActiveParam;
+  SmallVector<SourceKit::SignatureHelpResponse::Parameter, 1> Params;
+
+  SignatureInfo() {}
+};
+} // namespace
+
+/// \returns Array of parameters of \p VD accounting for implicitly curried
+/// instance methods.
+static ArrayRef<const ParamDecl *>
+getParameterArray(const ValueDecl *VD, bool IsImplicitlyCurried,
+                  const ParamDecl *&Scratch) {
+  if (!VD)
+    return {};
+
+  if (IsImplicitlyCurried) {
+    auto *FD = dyn_cast<AbstractFunctionDecl>(VD);
+    assert(FD && FD->hasImplicitSelfDecl());
+
+    Scratch = FD->getImplicitSelfDecl();
+    return ArrayRef(&Scratch, 1);
+  }
+
+  if (auto *ParamList = VD->getParameterList())
+    return ParamList->getArray();
+
+  return {};
+}
+
+static CodeCompletionString *createSignatureString(
+    llvm::BumpPtrAllocator &Allocator, ValueDecl *FD, AnyFunctionType *AFT,
+    const DeclContext *DC, GenericSignature GenericSig, bool IsSubscript,
+    bool IsMember, bool IsImplicitlyCurried, bool IsSecondApply) {
+  CodeCompletionStringBuilder StringBuilder(
+      Allocator, /*AnnotateResults=*/false,
+      /*UnderscoreEmptyArgumentLabel=*/!IsSubscript,
+      /*FullParameterFlags=*/true);
+
+  DeclBaseName BaseName;
+
+  if (!IsSecondApply && FD) {
+    BaseName = FD->getBaseName();
+  } else if (IsSubscript) {
+    BaseName = DeclBaseName::createSubscript();
+  }
+
+  if (!BaseName.empty())
+    StringBuilder.addValueBaseName(BaseName, IsMember);
+
+  StringBuilder.addLeftParen();
+
+  const ParamDecl *ParamScratch;
+  StringBuilder.addCallArgumentPatterns(
+      AFT->getParams(),
+      getParameterArray(FD, IsImplicitlyCurried, ParamScratch), DC, GenericSig,
+      DefaultArgumentOutputMode::All, /*includeDefaultValues=*/true);
+
+  StringBuilder.addRightParen();
+
+  if (!IsImplicitlyCurried) {
+    // For a second apply, we don't pass the declaration to avoid adding
+    // incorrect rethrows and reasync which are only usable in a single apply.
+    StringBuilder.addEffectsSpecifiers(
+        AFT,
+        /*AFD=*/IsSecondApply ? nullptr
+                              : dyn_cast_or_null<AbstractFunctionDecl>(FD));
+  }
+
+  if (FD && FD->isImplicitlyUnwrappedOptional())
+    StringBuilder.addTypeAnnotationForImplicitlyUnwrappedOptional(
+        AFT->getResult(), DC, GenericSig);
+  else
+    StringBuilder.addTypeAnnotation(AFT->getResult(), DC, GenericSig);
```

Comment:
Nit: Curly braces for `if-else`

```suggestion
  if (FD && FD->isImplicitlyUnwrappedOptional()) {
    StringBuilder.addTypeAnnotationForImplicitlyUnwrappedOptional(
        AFT->getResult(), DC, GenericSig);
  } else {
    StringBuilder.addTypeAnnotation(AFT->getResult(), DC, GenericSig);
  }
```